File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        
        maxSoFar = nums[0]
        minSoFar = nums[0]
        result = maxSoFar
        for i in range(1, len(nums)):
            curr = nums[i]
            tempMax = max(maxSoFar * curr, minSoFar * curr, curr)
            minSoFar = min(maxSoFar * curr, minSoFar * curr, curr)
            maxSoFar = tempMax
            result = max(result, maxSoFar)
        return result

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int minSoFar = nums[0];
        int maxSoFar = nums[0];
        int ans = maxSoFar;
        for (int i = 1; i < nums.length; i++) {
            int curr = nums[i];
            int temp = Math.max(curr, Math.max(curr * maxSoFar, curr * minSoFar));
            minSoFar = Math.min(curr, Math.min(curr * maxSoFar, curr * minSoFar));
            maxSoFar = temp;
            ans = Math.max(ans, maxSoFar);
        }
        return ans;
    }
}